src/ethereum/prague/fork.py

```diff
@@ -113,44 +118,57 @@ def apply_fork(old: BlockChain) -> BlockChain:
+def process_block_hash_history(chain: BlockChain, block: Block) -> None:
     """
-    Obtain the list of hashes of the previous 256 blocks in order of
-    increasing block number.
-
-    This function will return less hashes for the first 256 blocks.
-
-    The ``BLOCKHASH`` opcode needs to access the latest hashes on the chain,
-    therefore this function retrieves them.
+    Update the block hash history before block execution.
+    Introduced in EIP-2935. https://eips.ethereum.org/EIPS/eip-2935
 
     Parameters
     ----------
     chain :
         History and current state.
-
-    Returns
-    -------
-    recent_block_hashes : `List[Hash32]`
-        Hashes of the recent 256 blocks in order of increasing block number.
+    block :
+        Block to be executed.
     """
-    recent_blocks = chain.blocks[-255:]
-    # TODO: This function has not been tested rigorously
-    if len(recent_blocks) == 0:
-        return []
-
-    recent_block_hashes = []
-
-    for block in recent_blocks:
-        prev_block_hash = block.header.parent_hash
-        recent_block_hashes.append(prev_block_hash)
+    # Calling this function on blocks from previous forks is a bug
+    # since it determines whether `block` is the fork block
+    # by looking at the parent block's timestamp.
 
-    # We are computing the hash only for the most recent block and not for
-    # the rest of the blocks as they have successors which have the hash of
-    # the current block as parent hash.
-    most_recent_block_hash = keccak256(rlp.encode(recent_blocks[-1].header))
-    recent_block_hashes.append(most_recent_block_hash)
+    # The following ensures that `block` belongs to this fork.
+    assert FORK_CRITERIA.check(
+        block.header.number,
+        block.header.timestamp,
+    )
 
-    return recent_block_hashes
+    if FORK_CRITERIA.check(
+        chain.blocks[-1].header.number,
+        chain.blocks[-1].header.timestamp,
+    ):
+        # Set the hash for only one block
+        set_storage(
+            chain.state,
+            HISTORY_STORAGE_ADDRESS,
+            ((block.header.number - 1) % HISTORY_SERVE_WINDOW).to_be_bytes32(),
+            U256.from_be_bytes(block.header.parent_hash),
+        )
+    else:
+        # Parent timestamp is less than fork timestamp. Can only happen
+        # at fork block. Set hashes for the entire HISTORY_SERVE_WINDOW
+        current_block = block
+        for i in range(HISTORY_SERVE_WINDOW - 1):
+            if current_block.header.number == 0:
+                break
+
+            set_storage(
+                chain.state,
+                HISTORY_STORAGE_ADDRESS,
+                (
+                    (current_block.header.number - 1) % HISTORY_SERVE_WINDOW
+                ).to_be_bytes32(),
+                U256.from_be_bytes(current_block.header.parent_hash),
+            )
+            # Set the earlier block as the current block
+            current_block = chain.blocks[-1 - i]
```

Comment:
Wait, shouldn't this just be in the `apply_fork` function? Sorry I didn't catch that earlier!

If it can't be in the `apply_fork` function, we should add a comment explaining why.

Reply:
Updated